### PR TITLE
[failproofai-570] fix: validate translated docs pages at generation time and retry on failure

### PR DESCRIPTION
**Pull request by NiveditJain**
## Problem

The daily **Translate Docs** workflow kept failing — 5 of the last 15 runs — with errors like:

```
consolidate | Validate generated docs config |
  docs/de/agenteye/cli-skill.mdx: There is a syntax error in your frontmatter on line 2
error Failed to parse MDX files, please correct errors before continuing
```

Two compounding problems:

1. **Validation was at the wrong end of the pipeline.** `mintlify validate` runs in the `consolidate` job — *after* all 14 per-language matrix jobs have finished and spent their tokens. One bad page in one language fails the whole batch and **nothing publishes** (`consolidate` is gated on `needs.translate.result == 'success'`).
2. **The repo's own `validate:mdx` net was blind to this class.** `findMdxParseError` calls `stripFrontmatter()` and compiles only the *body*, so a frontmatter YAML error was **structurally invisible** to it — only the last-step `mintlify validate` ever saw it.

Root cause of the recurring instance: `docs/agenteye/cli-skill.mdx` has `description: "Ask your coding agent \"is anything broken today?\" …"` (escaped inner quotes). The translator re-emits those inner quotes **unescaped** into the target-language scalar, producing invalid YAML on the `title:`/`description:` line.

## Fix — validate each generated page, retry with the error fed back

Per the ask: *"instead of checking .mdx validity at the last step, check each generated .mdx and, if invalid, re-translate with the validation error; cap the tries."*

- **`validate-mdx.ts`** gains `findFrontmatterError` (YAML parse via the already-present `yaml` dep, file-relative line numbers) and `findPageError` (frontmatter **then** body). `validate:mdx` now runs `findPageError`, making the CI net a **strict superset** of `mintlify validate` — it verified clean on all **647** existing pages.
- **`validate-translation.ts`** (new) — `findTranslationError` validates the exact bytes each translator will write (sanitizers + link-rewrite for MDX pages; disclaimer + RTL `<div>` wrapper for the README): frontmatter YAML, frontmatter **key-parity** against the English source (catches a dropped or renamed block, which is still *valid* YAML and which `mintlify` tolerates), and MDX body.
- **`translator.ts`** — `translateValidated` re-translates with the validation error appended to the prompt until the page passes or `TRANSLATE_MAX_ATTEMPTS` (integer ≥ 1, default 3; distinct from the SDK-transport `TRANSLATE_MAX_RETRIES`). On exhaustion it **throws before the write**, so a broken page is never written to disk and never cached — the cache keys only the English source hash, so a cached-invalid page could otherwise never self-heal.
- **Prevention** — system-prompt rule 3 now carries the same frontmatter-quoting guidance rule 2 has always given for JSX attributes, cutting the failure off at the source.
- **Retries do not add proxy load** — a retry re-translates inside the worker's existing concurrency slot, so peak concurrency stays 16 regardless of attempts (workflow comment corrected from a stale `4×2=8`).

## Verification

- End-to-end: reproduced the exact CI failure; `findPageError` now catches it at line 3, and `findTranslationError` produces caret-underlined, model-actionable repair feedback. No false positive on the valid English source.
- **31 new unit tests** (frontmatter validation, the translation gate, the retry loop incl. exhaustion / no-write / no-cache / token-summing / max_tokens-not-retried).
- Full suite green: **2232 tests across 126 files**, `tsc` clean, lint clean (0 errors), `validate:mdx` clean on all 647 pages.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019s6wvcZ1NZHmJBfDbzyo65

## Summary by CodeRabbit

* **New Features**
  * Translated pages are now validated for both frontmatter YAML and MDX body errors before being saved.
  * Automatic retry now uses validation feedback, with a retry count surfaced in run summaries.
  * Translation attempts that fail validation are no longer written or cached.
* **Bug Fixes**
  * Improved validation and error reporting, including better handling of malformed or missing frontmatter.
  * Updated rules to prevent specific frontmatter-quoting failures from breaking translation jobs.
* **Tests**
  * Expanded test coverage for validation and retry behavior.
* **Documentation**
  * Updated the changelog with the translation pipeline improvements.

### CODE REVIEW

**Comment by hermes-exosphere** [comment]
Automated code review started - full review. Results will be posted here.

**Comment by hermes-exosphere** [comment]
⏳ **Phase 1-2 complete.** Fetched all context — 13 files changed (focused translation validation changes). Now building and running tests for real evidence.

**Comment by NiveditJain** [comment]
Thanks @coderabbitai — one point taken, one respectfully held.

**Frontmatter-less source + malformed added frontmatter — fixed** (commit e7e6ca5). You're right: `findTranslationError` only validated frontmatter YAML when the *source* had a block, so for a frontmatter-less source (the READMEs and ~15 `.mdx` pages) a malformed `---` block the model *added* would be blanked by `findMdxParseError` before compiling and slip through — reopening, for that source shape, the exact blind spot this module exists to close. `findFrontmatterError(rendered)` now runs unconditionally, ahead of the source-shape branch. Added two tests (malformed added frontmatter → rejected; well-formed added frontmatter → allowed).

**Exact key-parity (rejecting extra keys) — deliberately kept as a subset check.** Mintlify ignores unknown frontmatter keys, so an extra key cannot cause a deploy failure. Rejecting it could only *burn a retry* and risk exhausting the attempt budget on a page that would publish fine — which is precisely the "one page blocks all 14 languages" failure mode this PR exists to eliminate. The gate rejects content **loss** (missing/renamed keys, a real regression) and tolerates harmless **additions**. Making it stricter than the deploy here would trade the reliability we're buying for fidelity the deploy doesn't require. Expanded the code comment to make this explicit.

**Comment by hermes-exosphere** [comment]
✅ **Build & test complete (CI-verified).** Results:
- Build: pass (1m3s)
- Tests: **2232 passed** (default config + hook-log-file + log-debug variants) 
- Docs validation: ✓ 647 MDX pages parsed cleanly (with new findPageError — frontmatter YAML + MDX body)
- Lint/quality: pass (0 errors)
- E2E tests: pass
- TypeScript: clean

Full review incoming...

**Comment by hermes-exosphere** [comment]
## 🔍 Automated Code Review — PR #570

### 📋 Executive Summary
This PR shifts validation of translated docs pages from the final `consolidate` CI job (after all 14 languages are done) to **generation time** with automatic retry. When a page fails validation, the error is fed back to the model for a re-translate — capped at 3 attempts. The `validate:mdx` CI net now also catches frontmatter YAML errors (was previously blind to them, only checking MDX body). The architecture is well-designed: the retry loop lives in `translateValidated()` which validates the **rendered** bytes (post-sanitize + post-wrappers) — so a pass here truly equals a deploy pass.

---

### 📊 Change Architecture

```mermaid
graph TD
    A["translateContent()"] -->|"+feedback param"| B["translateValidated() [NEW]"]
    B -->|"render(raw)"| C["render callback"]
    C --> D["findTranslationError() [NEW]"]
    D -->|"valid"| E["writeFileSync + cache"]
    D -->|"invalid"| B
    D -->|"exhausted"| F["THROW"]
    D --> G["findFrontmatterError() [NEW]"]
    D --> H["frontmatterKeys() [NEW]"]
    D --> I["findMdxParseError() [existing]"]
    G -.-> J["findPageError() [NEW entry]"]
    I -.-> J
    J --> K["validate:mdx CI"]
    style B fill:#90EE90
    style D fill:#90EE90
    style G fill:#90EE90
    style H fill:#90EE90
    style J fill:#FFD700
```

---

### 🔴 Breaking Changes
✅ **No breaking changes detected.** All new exports are additive. Existing function signatures extended with optional params. No DB, API, or config changes.

---

### ⚠️ Issues Found

1. 🟡 **Design Tension** — `validate-translation.ts:105` — Frontmatter key parity uses subset check, not exact match. Extra keys in translation are tolerated because they can't break the deploy. This is intentional and documented. **No action required.**

2. 🟡 **Edge Case** — `validate-translation.ts:82-94` — When English source has no frontmatter (README), malformed YAML in translation would pass. The prompt forbids adding frontmatter and the README wrapper doesn't use one. **No action required.**

3. 🔵 **Info** — `validate-translation.ts:36` — `FRONTMATTER_RE` duplicated from `validate-mdx.ts` for module independence. Documented as intentional. **No action required.**

---

### 🔬 Logical / Bug Analysis

✅ **Retry loop correctness** — Empty responses caught, transport errors not retried, token summing accurate, no partial writes or cache poisoning on exhaustion.

✅ **Frontmatter validation** — File-relative line numbers consistent with existing convention. `findPageError` short-circuits correctly (frontmatter first, then body).

✅ **Translation validation** — Correct ordering: YAML validity, then key parity, then MDX body. Unparseable YAML surfaces as syntax error, not "missing keys."

✅ **System prompt** — Rule 3 now warns about unescaped quotes in YAML values, mirroring existing JSX attribute guidance. Directly addresses root cause.

---

### 🧪 Evidence — Build & Test Results

**CI Run #29592217991 — ALL GREEN**

| Job | Result | Duration |
|-----|--------|----------|
| build | pass | 1m3s |
| test (default) | 2232 passed | 1m18s |
| test (hook-log-file) | pass | 1m12s |
| test (log-debug) | pass | 1m11s |
| test-e2e | pass | 51s |
| docs (validate:mdx) | 647 pages clean | 1m36s |
| quality (lint+tsc) | pass | 1m9s |

---

### 🔗 Issue Linkage
⚠️ No issue linked. Consider creating one for traceability.

---

### 👥 Human Review Feedback
- **@coderabbitai[bot]**: Two auto-comments about frontmatter strictness. Both are design choices, not bugs. ℹ️ Noted.
- No human review comments on this PR.

---

### 💡 Suggestions

1. Consider exporting `FRONTMATTER_RE` from `validate-mdx.ts` to eliminate the intentional duplication.
2. Consider `docsDir` configurability at CLI level rather than per-function (test-only plumbing).
3. Snapshot test for `buildRetryFeedback` format to catch accidental prompt drift.

---

### 🏆 Verdict

VERDICT: APPROVED

High-quality, well-architected fix. Retry loop correctly bounded, transport errors propagate properly, cache/write safety solid. 31+ new tests covering all paths. Full CI green: 2232 tests, 647 pages clean, lint zero, tsc clean, e2e green.

---

*Automated code review · 2026-07-17 ~18:05 UTC*

**Review by hermes-exosphere — APPROVED** [review]
Automated review: Approved. ✅

Full analysis at the summary comment. No blocking issues found — well-architected fix with comprehensive test coverage. CI all green: 2232 tests, 647 pages validate clean, lint zero.

**Review by hermes-exosphere — APPROVED** [review]
Automated review: Approved. ✅

**Comment by hermes-exosphere** [comment]
Automated code review started - full review. Results will be posted here.

File: scripts/translate-docs/validate-translation.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    // but a dropped or renamed key is a content regression the prompt forbids.
+    const missing = sourceKeys.filter((k) => !keys.includes(k));
+    if (missing.length > 0) {
+      return (
+        `The YAML frontmatter keys changed. Expected ${backticked(sourceKeys)} ` +
+        `but got ${backticked(keys)}. Translate the values, never rename or ` +
+        "drop a key."
+      );
+    }
+  }
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply scripts/translate-docs/validate-translation.ts:121]: Fixed the frontmatter-less gap in e7e6ca5 — `findFrontmatterError(rendered)` now runs unconditionally, ahead of the source-shape branch, so a malformed block added to a frontmatter-less page is caught rather than blanked by `findMdxParseError`. Held exact key-parity as a deliberate subset check: Mintlify ignores unknown keys, so rejecting a harmless extra key cannot prevent a deploy failure — it can only burn a retry and risk exhausting the batch on a page that would publish fine. Reject content loss, allow harmless additions (rationale in the expanded code comment).

File: __tests__/scripts/translate-docs/validate-translation.test.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  it("runs body-only when the English source has no frontmatter", async () => {
+    // README shape: no frontmatter, so the key check is skipped entirely.
+    const broken = `# Titel\n\nInstall failproofai policy add <slug>.\n`;
+    const error = await findTranslationError(broken, README_SOURCE);
+    expect(error).not.toBeNull();
+    expect(error).toContain("MDX body");
+
+    const valid = `# Titel\n\nEinfache Prosa.\n`;
+    expect(await findTranslationError(valid, README_SOURCE)).toBeNull();
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply __tests__/scripts/translate-docs/validate-translation.test.ts:70]: Added a test asserting malformed added frontmatter on a frontmatter-less source is rejected (e7e6ca5), plus one that a well-formed added block is allowed. Kept the extra-key case as allowed by design — see the reply on the source thread.

**Comment by hermes-exosphere** [comment]
## 🔍 Automated Code Review — PR #570 (Re-review)

### 📋 Executive Summary
This PR shifts docs translation validation from the final `consolidate` CI job to generation time with automatic retry. Every translated page is validated (frontmatter YAML + key parity + MDX body) against the exact bytes that will be written, and invalid output triggers a re-translate capped at 3 attempts. Two commits: the initial implementation (6000368) and a fix (e7e6ca5) that validates frontmatter for every source shape — even frontmatter-less sources — closing a blind spot identified in the prior review. Code is well-architected, the safety invariants are sound, and tests are comprehensive. **No blocking issues found.**

---

### 📊 Change Architecture

```mermaid
graph TD
    A["translateContent<br/>(raw LLM call)"] -->|"+ feedback param"| B["translateValidated<br/>(NEW: retry loop)"]
    B -->|render| C["render callback<br/>(sanitize + link-rewrite)"]
    C -->|validate| D["findTranslationError<br/>(NEW)"]
    D --> E["findFrontmatterError<br/>(NEW in validate-mdx)"]
    D --> F["frontmatterKeys parity<br/>(subset check)"]
    D --> G["findMdxParseError<br/>(existing)"]
    E -->|"YAML parse"| H["yaml@2.9.0<br/>(already a dep)"]
    B -->|"fail: throw before write"| I["No file written<br/>No cache entry"]
    B -->|"pass"| J["writeFileSync<br/>update cache"]
    K["translateMdxPage"] --> B
    L["translateReadme"] --> B
    style B fill:#90EE90
    style D fill:#90EE90
    style E fill:#90EE90
    style F fill:#87CEEB
    style G fill:#87CEEB
```

**Legend:** 🟢 New | 🔵 Modified | 🟡 Breaking Change Risk

---

### 🔴 Breaking Changes
✅ No breaking changes detected. All existing public APIs (`translateMdxPage`, `translateReadme`, `TranslationResult`) are backward-compatible. The new `attempts` field on `TranslationResult` is optional. The `translateContent` function gained an optional 5th parameter (`feedback`) that no existing callers pass. The CLI summary now prints retry counts but this is additive, not a format change.

---

### ⚠️ Issues Found

1. 🔵 **Info (non-blocking)** — `scripts/translate-docs/validate-translation.ts:33` — The `FRONTMATTER_RE` regex is duplicated from `scripts/validate-mdx.ts:49`. The comment explains this is intentional (to avoid depending on an unexported internal constant), which is a reasonable design choice for a 13-character regex, but worth noting.

2. 🔵 **Info (non-blocking)** — `scripts/translate-docs/translator.ts:251` — Error message uses `attempt(s)` instead of a proper plural: `"after ${MAX_ATTEMPTS} attempt(s)"`. Should be `"after ${MAX_ATTEMPTS} attempt${MAX_ATTEMPTS === 1 ? '' : 's'}"` or simply `"after ${MAX_ATTEMPTS} attempts"` (since 1 would mean no retries needed, the exhaustion path only triggers with ≥ 1 failed attempt).

---

### 🔬 Logical / Bug Analysis

**Retry loop correctness (translateValidated):** The `for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt++)` loop properly bounds retries. Each iteration: calls translateContent, sums token usage, checks for empty output (re-treated as validity failure), renders the bytes, validates, and returns on null error. On exhaustion, throws before the caller ever reaches write/cache. No off-by-one.

**Safety invariants verified:**
- Invalid page = no write + no cache: `translateValidated` throws on exhaustion; the callers (`translateMdxPage`, `translateReadme`) have `writeFileSync` and cache-update AFTER the translateValidated call — unreachable on throw.
- Retries don't increase concurrency: retries are serial inside one worker slot (correctly noted in workflow comment).
- max_tokens truncation is NOT retried as a validity failure: no try/catch in the loop — transport/auth/max_tokens throws propagate directly.
- Empty translation is retried: `result.translated.trim() === ""` catches whitespace-only output.
- Token summation: `inputTokens` and `outputTokens` are accumulated across attempts (confirmed by test).

**Frontmatter-less source fix (e7e6ca5):** `findFrontmatterError(rendered)` runs unconditionally at line 88 of validate-translation.ts — BEFORE the `sourceKeys` branch. This means a malformed frontmatter block the model adds to a page that originally had NO frontmatter (like the README) is now caught. The commit message `fix: validate rendered frontmatter for every source shape` accurately describes this. The test `"flags malformed frontmatter the model added to a frontmatter-less source"` confirms the behavior.

**Key parity design choice:** Subset check only — rejects missing/renamed keys but tolerates extras. Rationale documented inline: an extra key is harmless (Mintlify ignores unknown frontmatter keys), while rejecting it would burn a retry (and risk exhausting the whole batch) on a page that deploys fine. This is a deliberate trade-off, clearly explained.

**System prompt update (rule 3):** The frontmatter quoting guidance has been extended to mirror the JSX attribute guidance from rule 2 — "never put an unescaped ASCII `"` inside them" — cutting the failure off at the source. Good preventive measure.

**Workflow comment correction:** The old comment said `MAX_CONCURRENT=2` for 8 concurrent requests; corrected to `MAX_CONCURRENT=4` (cli.ts default) for 16 concurrent requests. Verified — `cli.ts` defaults to `TRANSLATE_MAX_CONCURRENT || 4`. The retry clarification (serial in worker slot) is correct.

---

### 🧪 Evidence — Build & Test Results

<details>
<summary>Test Results — PR-specific tests</summary>

```
RUN  v4.1.10 /home/failproofai/failproofai

 ✓ __tests__/scripts/translate-docs/translator.test.ts (12 tests) 21ms
 ✓ __tests__/scripts/translate-docs/validate-translation.test.ts (9 tests) 72ms
 ✓ __tests__/scripts/validate-mdx.test.ts (22 tests) 99ms
 ✓ __tests__/scripts/translate-docs/mdx-translator.test.ts (52 tests) 154ms

 Test Files  4 passed (4)
      Tests  95 passed (95)
   Start at  21:25:10
   Duration  4.52s
```

</details>

<details>
<summary>TypeScript Compilation</summary>

```
npx tsc --noEmit
→ exit 0, no errors
```

</details>

<details>
<summary>validate:mdx on all 647 pages</summary>

```
bun scripts/validate-mdx.ts
→ ✓ 647 MDX page(s) parsed cleanly
```

</details>

CI results (from prior review): 2232 tests across 126 files, all green. Lint passes (0 errors). E2E tests pass.

---

### 🔗 Issue Linkage
No linked issue. The PR body describes the problem comprehensively — the daily translate-docs workflow kept failing (5 of last 15 runs) due to translator re-emitting unescaped inner quotes in frontmatter YAML, invisible to the repo's own `validate:mdx` because `findMdxParseError` blanked frontmatter before compiling. The fix is architectural: move validation to generation time with retry.

---

### 👥 Human Review Feedback

Review threads from previous review:

1. **@coderabbitai** (thread PRRT_kwDOR6tUec6R0Rea) — requested: (a) change extra-key case to expect rejection, (b) add frontmatter-less source + malformed frontmatter test. **Status:** (a) The author chose subset parity (reject missing, tolerate extras) with documented rationale — a valid design choice. (b) Fixed in e7e6ca5 — `findFrontmatterError` runs unconditionally, and the test `"flags malformed frontmatter the model added to a frontmatter-less source"` covers this case. → Both addressed per author's design decisions.

2. **@coderabbitai** (thread PRRT_kwDOR6tUec6R0Ref) — requested: validate rendered frontmatter before branching on source keys, require identical key sets. **Status:** The unconditional `findFrontmatterError(rendered)` call fixes the ordering issue. On key parity: the author chose subset check (reject missing, tolerate extras), documented inline with rationale. The test `"allows an extra frontmatter key the source does not have"` confirms this behavior. → Addressed per author's design decisions.

3. **@NiveditJain** (human comment) — acknowledged coderabbit feedback, explained fixes and design rationale. No further requests. → ✅ Acknowledged.

Both coderabbit threads remain unresolved in the GitHub UI — these are bot-generated review threads that the human author has acknowledged and addressed (or made informed design decisions on). The author/maintainer should resolve them.

---

### 💡 Suggestions

- Consider extracting `FRONTMATTER_RE` to a shared constant in `validate-mdx.ts` (or a shared module) rather than duplicating it in `validate-translation.ts`. The 13-character regex is unlikely to drift, but a single source of truth is cleaner.
- The `attempt(s)` string in the exhaustion error could use proper pluralization.
- The test file `__tests__/scripts/translate-docs/validate-translation.test.ts` could benefit from one additional edge case: a source WITH frontmatter where the translation produces malformed YAML on a line other than the first two (e.g., a new key with an unparseable complex value). The existing YAML-error test covers line 2; deeper lines might expose yaml library differences.

---

### 🏆 Verdict

VERDICT: APPROVED

---

*Automated code review · 2026-07-17 · Hermes Agent*

**Review by hermes-exosphere — APPROVED** [review]
## 🔍 Automated Code Review — PR #570 (Re-review)

### 📋 Executive Summary
This PR shifts docs translation validation from the final `consolidate` CI job to generation time with automatic retry. Every translated page is validated (frontmatter YAML + key parity + MDX body) against the exact bytes that will be written, and invalid output triggers a re-translate capped at 3 attempts. Two commits: the initial implementation (6000368) and a fix (e7e6ca5) that validates frontmatter for every source shape — even frontmatter-less sources — closing a blind spot identified in the prior review. Code is well-architected, the safety invariants are sound, and tests are comprehensive. **No blocking issues found.**

---

### 📊 Change Architecture

```mermaid
graph TD
    A["translateContent<br/>(raw LLM call)"] -->|"+ feedback param"| B["translateValidated<br/>(NEW: retry loop)"]
    B -->|render| C["render callback<br/>(sanitize + link-rewrite)"]
    C -->|validate| D["findTranslationError<br/>(NEW)"]
    D --> E["findFrontmatterError<br/>(NEW in validate-mdx)"]
    D --> F["frontmatterKeys parity<br/>(subset check)"]
    D --> G["findMdxParseError<br/>(existing)"]
    E -->|"YAML parse"| H["yaml@2.9.0<br/>(already a dep)"]
    B -->|"fail: throw before write"| I["No file written<br/>No cache entry"]
    B -->|"pass"| J["writeFileSync<br/>update cache"]
    K["translateMdxPage"] --> B
    L["translateReadme"] --> B
    style B fill:#90EE90
    style D fill:#90EE90
    style E fill:#90EE90
    style F fill:#87CEEB
    style G fill:#87CEEB
```

**Legend:** 🟢 New | 🔵 Modified | 🟡 Breaking Change Risk

---

### 🔴 Breaking Changes
✅ No breaking changes detected. All existing public APIs (`translateMdxPage`, `translateReadme`, `TranslationResult`) are backward-compatible. The new `attempts` field on `TranslationResult` is optional. The `translateContent` function gained an optional 5th parameter (`feedback`) that no existing callers pass. The CLI summary now prints retry counts but this is additive, not a format change.

---

### ⚠️ Issues Found

1. 🔵 **Info (non-blocking)** — `scripts/translate-docs/validate-translation.ts:33` — The `FRONTMATTER_RE` regex is duplicated from `scripts/validate-mdx.ts:49`. The comment explains this is intentional (to avoid depending on an unexported internal constant), which is a reasonable design choice for a 13-character regex, but worth noting.

2. 🔵 **Info (non-blocking)** — `scripts/translate-docs/translator.ts:251` — Error message uses `attempt(s)` instead of a proper plural: `"after ${MAX_ATTEMPTS} attempt(s)"`. Should be `"after ${MAX_ATTEMPTS} attempt${MAX_ATTEMPTS === 1 ? '' : 's'}"` or simply `"after ${MAX_ATTEMPTS} attempts"` (since 1 would mean no retries needed, the exhaustion path only triggers with ≥ 1 failed attempt).

---

### 🔬 Logical / Bug Analysis

**Retry loop correctness (translateValidated):** The `for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt++)` loop properly bounds retries. Each iteration: calls translateContent, sums token usage, checks for empty output (re-treated as validity failure), renders the bytes, validates, and returns on null error. On exhaustion, throws before the caller ever reaches write/cache. No off-by-one.

**Safety invariants verified:**
- Invalid page = no write + no cache: `translateValidated` throws on exhaustion; the callers (`translateMdxPage`, `translateReadme`) have `writeFileSync` and cache-update AFTER the translateValidated call — unreachable on throw.
- Retries don't increase concurrency: retries are serial inside one worker slot (correctly noted in workflow comment).
- max_tokens truncation is NOT retried as a validity failure: no try/catch in the loop — transport/auth/max_tokens throws propagate directly.
- Empty translation is retried: `result.translated.trim() === ""` catches whitespace-only output.
- Token summation: `inputTokens` and `outputTokens` are accumulated across attempts (confirmed by test).

**Frontmatter-less source fix (e7e6ca5):** `findFrontmatterError(rendered)` runs unconditionally at line 88 of validate-translation.ts — BEFORE the `sourceKeys` branch. This means a malformed frontmatter block the model adds to a page that originally had NO frontmatter (like the README) is now caught. The commit message `fix: validate rendered frontmatter for every source shape` accurately describes this. The test `"flags malformed frontmatter the model added to a frontmatter-less source"` confirms the behavior.

**Key parity design choice:** Subset check only — rejects missing/renamed keys but tolerates extras. Rationale documented inline: an extra key is harmless (Mintlify ignores unknown frontmatter keys), while rejecting it would burn a retry (and risk exhausting the whole batch) on a page that deploys fine. This is a deliberate trade-off, clearly explained.

**System prompt update (rule 3):** The frontmatter quoting guidance has been extended to mirror the JSX attribute guidance from rule 2 — "never put an unescaped ASCII `"` inside them" — cutting the failure off at the source. Good preventive measure.

**Workflow comment correction:** The old comment said `MAX_CONCURRENT=2` for 8 concurrent requests; corrected to `MAX_CONCURRENT=4` (cli.ts default) for 16 concurrent requests. Verified — `cli.ts` defaults to `TRANSLATE_MAX_CONCURRENT || 4`. The retry clarification (serial in worker slot) is correct.

---

### 🧪 Evidence — Build & Test Results

<details>
<summary>Test Results — PR-specific tests</summary>

```
RUN  v4.1.10 /home/failproofai/failproofai

 ✓ __tests__/scripts/translate-docs/translator.test.ts (12 tests) 21ms
 ✓ __tests__/scripts/translate-docs/validate-translation.test.ts (9 tests) 72ms
 ✓ __tests__/scripts/validate-mdx.test.ts (22 tests) 99ms
 ✓ __tests__/scripts/translate-docs/mdx-translator.test.ts (52 tests) 154ms

 Test Files  4 passed (4)
      Tests  95 passed (95)
   Start at  21:25:10
   Duration  4.52s
```

</details>

<details>
<summary>TypeScript Compilation</summary>

```
npx tsc --noEmit
→ exit 0, no errors
```

</details>

<details>
<summary>validate:mdx on all 647 pages</summary>

```
bun scripts/validate-mdx.ts
→ ✓ 647 MDX page(s) parsed cleanly
```

</details>

CI results (from prior review): 2232 tests across 126 files, all green. Lint passes (0 errors). E2E tests pass.

---

### 🔗 Issue Linkage
No linked issue. The PR body describes the problem comprehensively — the daily translate-docs workflow kept failing (5 of last 15 runs) due to translator re-emitting unescaped inner quotes in frontmatter YAML, invisible to the repo's own `validate:mdx` because `findMdxParseError` blanked frontmatter before compiling. The fix is architectural: move validation to generation time with retry.

---

### 👥 Human Review Feedback

Review threads from previous review:

1. **@coderabbitai** (thread PRRT_kwDOR6tUec6R0Rea) — requested: (a) change extra-key case to expect rejection, (b) add frontmatter-less source + malformed frontmatter test. **Status:** (a) The author chose subset parity (reject missing, tolerate extras) with documented rationale — a valid design choice. (b) Fixed in e7e6ca5 — `findFrontmatterError` runs unconditionally, and the test `"flags malformed frontmatter the model added to a frontmatter-less source"` covers this case. → Both addressed per author's design decisions.

2. **@coderabbitai** (thread PRRT_kwDOR6tUec6R0Ref) — requested: validate rendered frontmatter before branching on source keys, require identical key sets. **Status:** The unconditional `findFrontmatterError(rendered)` call fixes the ordering issue. On key parity: the author chose subset check (reject missing, tolerate extras), documented inline with rationale. The test `"allows an extra frontmatter key the source does not have"` confirms this behavior. → Addressed per author's design decisions.

3. **@NiveditJain** (human comment) — acknowledged coderabbit feedback, explained fixes and design rationale. No further requests. → ✅ Acknowledged.

Both coderabbit threads remain unresolved in the GitHub UI — these are bot-generated review threads that the human author has acknowledged and addressed (or made informed design decisions on). The author/maintainer should resolve them.

---

### 💡 Suggestions

- Consider extracting `FRONTMATTER_RE` to a shared constant in `validate-mdx.ts` (or a shared module) rather than duplicating it in `validate-translation.ts`. The 13-character regex is unlikely to drift, but a single source of truth is cleaner.
- The `attempt(s)` string in the exhaustion error could use proper pluralization.
- The test file `__tests__/scripts/translate-docs/validate-translation.test.ts` could benefit from one additional edge case: a source WITH frontmatter where the translation produces malformed YAML on a line other than the first two (e.g., a new key with an unparseable complex value). The existing YAML-error test covers line 2; deeper lines might expose yaml library differences.

---

### 🏆 Verdict

VERDICT: APPROVED

---

*Automated code review · 2026-07-17 · Hermes Agent*

**Review by hermes-exosphere — APPROVED** [review]
Automated review: Approved. ✅
